Added new tool - expressive-module (#8 and #9)

michalbundyra opened the pull request:
New tool `vendor/bin/expressive-module`

Tool provide following commands:
- `create` - create skeleton of the expressive module and register it
- `register` - register module in configuration and composer autoloading
- `deregister` - deregister module from configuration and composer autoloading

Tool accept following arguments:
- `--composer|-c` path to composer binary
- `--modules-path|-p` - path to modules directory

TODO:
- [x] tests
  - [x] actions
    - [x] create
    - [x] register
    - [x] deregister
  - [x] main command
  - [x] help
- [x] implement disable module's composer autoloading - will be nice to have it in [`zfcampus/zf-composer-autoloading`](https://github.com/zfcampus/zf-composer-autoloading) as there is also registering module in composer.json - done in https://github.com/zfcampus/zf-composer-autoloading/issues/8
- [x] would be nice to refactor [`zendframework/zend-component-installer`](https://github.com/zendframework/zend-component-installer) then we can have here less dependencies - now because we are using this library we require here `composer/composer` - done in https://github.com/zendframework/zend-component-installer/pull/34.

Resolves #8 and #9

michalbundyra committed:
Added new tool - expressive-module

Tool provide following commands:
- "create" - create skeleton of the expressive module and register it
- "register" - register module in configuration and composer autoloading
- "deregister" - deregister module from configuration and composer
  autoloading

Tool accept following arguments:
- "--composer|-c" path to composer binary
- "--modules-path|-p" - path to modules directory

weierophinney commented:
Syntax is perfect - exactly what I'd expect! Looking forward to reviewing the final iterations!

michalbundyra committed:
Used refactor version of zend-component-installer

Eliminated dependency of composer/composer package.
Now to inject/remove methods we don't need to pass
IOInterface.

michalbundyra committed:
PHPDocs fixed allowed param type

michalbundyra committed:
Removed unused import

michalbundyra committed:
Added command chains implementation

michalbundyra committed:
Removed redundant blank lines

michalbundyra committed:
Updated chmod for bin file (+x)

michalbundyra committed:
Suppressing output on checking composer binary

michalbundyra committed:
Updated to zf-composer-autoloading v2-dev with enable and disable commands

michalbundyra committed:
Method getCommand expects only first script argument

michalbundyra committed:
Suppress errors on checking composer binary

Redirect stderr to stdout 2>&1
Works also on Windows.

michalbundyra committed:
Command's process methods ca throw only module RuntimeException

Added catching RuntimeException from zf-composer-autoloading
and throwing module RuntimeException.
Now we have then consistent interface, and process method throws
only exception from the module.

Updated PHPDocs to reflect these changes.

michalbundyra committed:
Updated PHPDocs comment

michalbundyra committed:
Allow to use with zend-code 2.6.3

michalbundyra committed:
Added expressive-module Help tests

michalbundyra committed:
Updated to latest dev version of zfcampus/zf-composer-autoloading

Now commands from zfcampus/zf-composer-autoloading are invoked
via "process" method instead of "__invoke".
Also disabled moving module class for the module on enabling autoloading.

michalbundyra committed:
Construct command without moduleName and expect module name in process

michalbundyra committed:
Removed "chmod" property in Create command

michalbundyra committed:
Use constant from interface, not from injector

I've notice issue with constant and Mockery.
If we use Mockery to mock class, than we can't access to the constants.
This change uses constants from another class, so we have access
on mocking ConfigAggregatorInjector.

michalbundyra committed:
Added tests for commands "create", "register", "deregister"

michalbundyra committed:
Changed $projectDir property visibility to private

michalbundyra committed:
Updated arguments checks and messages

michalbundyra committed:
Added tests for Module\Command

michalbundyra commented:
@weierophinney It's ready for review, tests are failing but some of them are fixed in #10, so it should be merged first. Of course it needs also `zend-component-install 0.7.0` release and update requirements in `composer.json`.
What is missing here - messages on success. Now we don't have any messages, so it's confusing if it was done successfully or not. Just response code is 0 on success. I need to think about it a bit more how to do it nicely, or maybe you have an idea?

weierophinney commented:
@webimpress I've tagged zend-component-installer 0.7.0 at this time, and merged #10.

As for messages, how about:

- "Created module `<module name>` in `<path>`, and generated autoloading rules"
- "Registered autoloading rules and added configuration entry for module `<module>`"
- "Removed autoloading rules and configuration entries for module `<module>`"

michalbundyra committed:
Updated to released dependencies

- zendframework/zend-component-installer ^0.7
- zf-composer-autoloading ^2.0

michalbundyra committed:
Updated travis configuration to work with legacy deps

michalbundyra committed:
Added success messages

michalbundyra commented:
@weierophinney Rebased, used latest stable dependencies and added success messages as suggested. Ready for review.

weierophinney commented on composer.json:
```diff
@@ -40,7 +44,8 @@
         "bin/expressive-pipeline-from-config",
-        "bin/expressive-scan-for-error-middleware"
+        "bin/expressive-scan-for-error-middleware",
+        "bin/expressive-module"
```
These should be alphabetized.

weierophinney commented on src/Module/Command/Create.php:
```diff
@@ -0,0 +1,150 @@
+/**
+ * The configuration provider for the %1$s module
+ *
+ * @see https://docs.zendframework.com/zend-component-installer/
```
Nice touch! Linking to the documentation will definitely assist new developers.

weierophinney commented on src/Module/Command/Create.php:
```diff
@@ -0,0 +1,150 @@
+     */
+    private function createDirectoryStructure($moduleName)
+    {
+        if (file_exists($this->modulePath)) {
```
Considering that `$modulePath` is calculated by `process()`, perhaps it should be passed as an argument to each of this method and `createConfigProvider()`, instead of memoized as a property? That would be one less state for the instance to track.

michalbundyra replied:
Changed - property removed, now method has two parameters.

weierophinney commented on src/Module/Help.php:
```diff
@@ -0,0 +1,75 @@
+<info>Commands:</info>
+
+  <info>help</info>          Display this help/usage message
+  <info>create</info>        Create skeleton of the expressive module and register it
```
`s/skeleton of/source tree for/`

weierophinney commented on src/Module/Help.php:
```diff
@@ -0,0 +1,75 @@
+
+  <info>help</info>          Display this help/usage message
+  <info>create</info>        Create skeleton of the expressive module and register it
+  <info>register</info>      Register module in configuration and composer autoloading
```
"Register module in application configuration, and enable autoloading of module via composer."

weierophinney commented on src/Module/Help.php:
```diff
@@ -0,0 +1,75 @@
+  <info>help</info>          Display this help/usage message
+  <info>create</info>        Create skeleton of the expressive module and register it
+  <info>register</info>      Register module in configuration and composer autoloading
+  <info>deregister</info>    Deregister module from configuration and composer autoloading
```
"Deregister module from application configuration, and disable autoloading of module via composer."

weierophinney commented on test/Module/Command/CreateTest.php:
```diff
@@ -0,0 +1,129 @@
+    {
+        $baseModulePath = sprintf('%s/my-modules/MyApp', $this->dir->url());
+
+        $mkdir = $this->getFunctionMock('Zend\Expressive\Tooling\Module\Command', 'mkdir');
```
I was not aware of `getFunctionMock()`! Awesome!

michalbundyra replied:
Please not this is _not_ from PHPUnit but from `php-mock` library. Integration with PHPUnit is via `php-mock/php-mock-phpunit`.

weierophinney replied:
Ah, okay - did you add that dependency, or does PHPUnit depend on it itself? Either way, it's a nice library to know about.

michalbundyra replied:
Yes, I added:
```json
"php-mock/php-mock-phpunit": "^2.0 || ^1.1.2",
```
There is also integration library with prophecy and mocker.
https://github.com/php-mock
I've chosen integration with PHPUnit as this is the main testing library. I know that many mocking we are doing recently using prophecy, so maybe better choice will be `php-mock-prophecy`?
Here is the library with the main functionality:
https://github.com/php-mock/php-mock

weierophinney commented on test/Module/Command/DeregisterTest.php:
```diff
@@ -0,0 +1,110 @@
+namespace ZendTest\Expressive\Tooling\Module\Command;
+
+use Mockery;
+use Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;
```
Is this what allows you to not need to specify `@runInSeparateProcess` for tests that overload classes via Mockery?

michalbundyra replied:
Hm... this is recommended in Mockery documentation:
http://docs.mockery.io/en/latest/reference/phpunit_integration.html

I was trying to remove `@runInSeparateProcess` but some tests on some versions (php versions/dependencies) were failing. As you can note in some tests we still have this annotation.

weierophinney replied:
I've somehow missed that documentation page previously, so, again, a nice thing to know about!

weierophinney reviewed (changes requested):
Looks fantastic, @webimpress ! There are a few minor changes I've suggested, and one larger one that, honestly, I can go either way on (whether or not to memorize `$modulePath`). Let me know how you want to proceed on that, and I can merge.

michalbundyra committed:
Reorder bin scripts in composer - alphabetica order

- per @weierophinney

michalbundyra committed:
Removed class property and pass modulePath as argument

- per @weierophinney

michalbundyra committed:
Updated help messages

- per @weierophinney

weierophinney reviewed (approved):
Changes look perfect! Thanks, @webimpress !
